### ISSUE TITLE
Different VAMPNET results from CPU/GPU training

### ISSUE THREAD

**Issue by yuxuanzhuang** [deeptime-ml/deeptime#220]
**Describe the bug**
I tried to run the ala2 notebook (https://github.com/deeptime-ml/deeptime-notebooks/blob/master/examples/ala2-example.ipynb) but ended up with quite different results with GPU v.s. CPU training. CPU had a much higher success rate and flat training curves compared to GPU. I am wondering if it is something common or if I had made any mistakes. 

**Results**
I tested with 10 individual runs with the same parameters as in the tutorial notebook.

- CPU
![cpu_training](https://user-images.githubusercontent.com/24636656/159483538-f0d9437b-0f94-4a09-bdee-9d2e7abaff54.png)
![cpu_state](https://user-images.githubusercontent.com/24636656/159484348-8ec2d277-b3b0-44d8-9247-33e06be938fa.png)

- GPU
![gpu_training](https://user-images.githubusercontent.com/24636656/159483561-1f12e2a6-735b-4f4f-ad03-d02227ccdae5.png)
![gpu_state](https://user-images.githubusercontent.com/24636656/159484533-395db6c0-1dc6-49f2-80fd-cc26912c835a.png)

**System**
CPU: AMD EPYC 7551
GPU: RTX A5000
System: Ubuntu 20.04.1
Python 3.9
torch 1.11.0+cu113
deeptime '0.4.1+8.g38b0158.dirty' (main branch)

**Comment by yuxuanzhuang** [comment]
I tested both GTX980Ti and RTX2080 with the same cuda/conda environment and they worked perfectly fine:) As I don't have the privilege to restart the system, I tried different nodes equipped with A5000 in our cluster and they failed consistently. I also tried `torch.cuda.empty_cache()` and it's not helping as well.

I will see if it is a compatibility issue with pytorch by testing out other deep learning benchmarks. I will get back to you if I find out what's wrong.

**Comment by clonker** [comment]
Hmmm. My best guess is the driver version that is not fully compatible (yet) with A5000s. But really I am not sure. Please do keep me posted! Thanks

**Comment by yuxuanzhuang** [comment]
Found the solution! The culprit is the TF32 tensor cores on new Ampere devices. I have to manually set `torch.backends.cuda.matmul.allow_tf32 = False` to increase the accuracy of eigensolver for Koopman. (https://pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices)

Related issue: https://github.com/pytorch/pytorch/issues/58434

**Comment by clonker** [comment]
That is interesting. I should put up a warning in the vampnets documentation that this might be required. Good detective work!

**Comment by yuxuanzhuang** [comment]
Does it make sense to set it to `False` in default in the codebase given it is obviously not only failing ala2? Or add it as a context decorator only to the `vampnet_loss` in case people gain any performance elsewhere with this setting? I can take a stab at it.

**Comment by yuxuanzhuang** [comment]
It turns out tf32 needs to be disabled throughout training (and validating).

Results after fixing.
<img width="580" alt="image" src="https://user-images.githubusercontent.com/24636656/159909388-5430e322-cfa5-483c-8df0-87eaaacfaa10.png">

**Comment by clonker** [comment]
This might be a general problem with applications where precision does matter... for example with SO(3) equivarient nets etc. I am curious to see how things evolve.

**Comment by davidgilbertson** [comment]
In case you missed it, this tf32 setting is now [false by default from PyTorch 1.12](https://github.com/pytorch/pytorch/releases/tag/v1.12.0).

**Comment by clonker** [comment]
Thanks @davidgilbertson, i missed this indeed!
